### PR TITLE
SSH timeout added

### LINKED ISSUE #49: Allow users to configure a session timeout for SSH tunnels
The admin endpoint hangs when the SSH tunnels uses an unreachable host.

### PR DESCRIPTION
**Pull request by ddcprg**
Fixes #49 

I'll be running functional tests today, please take a look in the meantime

### CODE REVIEW

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnel.java [hunk 1]
```diff
@@ -26,12 +26,14 @@
...
   private static final int DEFAULT_PORT = 22;
   private static final String DEFAULT_LOCALHOST = "localhost";
+  private static final int DEFAULT_TIMEOUT_MILLIS = 60000; // 1 minute
 
   private @NotBlank @TunnelRoute String route;
   private @Min(1) @Max(65535) int port = DEFAULT_PORT;
   private String localhost = DEFAULT_LOCALHOST;
   private @NotBlank String privateKeys;
   private @NotBlank String knownHosts;
+  private @Min(0) int timeout = DEFAULT_TIMEOUT_MILLIS;
```
Comment (massdosage) [inline waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnel.java]: I think we should either name this timeoutMillis to be clear what unit it is or we introduce two values here - one which is a TimeUnit and the other being the value.

File: waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java [hunk 2]
```diff
@@ -49,6 +50,15 @@ public void typical() {
     assertThat(violations.size(), is(0));
   }
 
+  @Test
+  public void infiniteTimeout() {
```
Comment (massdosage) [inline waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java]: We should document in the README that setting it to zero means no timeout.

File: waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java [hunk 3]
```diff
@@ -148,4 +158,13 @@ public void blankPrivateKey() {
     assertThat(violations.size(), is(1));
   }
 
+  @Test
+  public void negativeTunnel() {
```
Comment (massdosage) [inline waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java]: negativeTimeout

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SessionFactorySupplier.java [hunk 4]
```diff
@@ -31,15 +34,40 @@
...
+  private final int sshTimeout;
   private DefaultSessionFactory sessionFactory = null;
 
+  @Deprecated
   public SessionFactorySupplier(int sshPort, String knownHosts, List<String> identityKeys) {
+    this(sshPort, knownHosts, identityKeys, 0);
+  }
+
+  public SessionFactorySupplier(int sshPort, String knownHosts, List<String> identityKeys, int sshTimeout) {
+    Preconditions.checkArgument(0 <= sshPort && sshPort <= 65535, "Invalid SSH port number " + sshPort);
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SessionFactorySupplier.java]: Why do we have this check both here and in the TunnelHandler  but we only have the timeout check here and not in the TunnelHandler ?

Comment (ddcprg) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SessionFactorySupplier.java]: I'm improving the code, the old version does not check... the constructor is public

File: waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java (hunk 3 above)
Comment (ddcprg) [reply waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java]: yeah, typo.. will rename

File: waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java (hunk 2 above)
Comment (ddcprg) [reply waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnelTest.java]: will document

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnel.java (hunk 1 above)
Comment (ddcprg) [reply waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnel.java]: Even if the docs say the units is millis?

Comment (massdosage) [reply waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/MetastoreTunnel.java]: Yeah, I guess it's fine as long as its clearly documented.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SessionFactorySupplier.java (hunk 4 above)
Comment (massdosage) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SessionFactorySupplier.java]: OK

**Comment by ddcprg** [comment]
This just got uglier but it works. Please take a look again and I'll try to do a release tomorrow.

**Pull request merged**